packages/json-rpc-engine/src/v2/JsonRpcEngineV2.ts

```diff
@@ -0,0 +1,338 @@
+  middleware: NonEmptyArray<JsonRpcMiddleware<Request, Result | void>>;
+};
+
+type HandleOptions = {
+  context?: MiddlewareContext;
+};
+
+/**
+ * A JSON-RPC request and response processor.
+ *
+ * Give it a stack of middleware, pass it requests, and get back responses.
+ *
+ * @template Request - The type of request to handle.
+ * @template Result - The type of result to return.
+ *
+ * @example
+ * ```ts
+ * const engine = new JsonRpcEngineV2({
+ *   middleware,
+ * });
+ *
+ * try {
+ *   const result = await engine.handle(request);
+ *   // Handle result
+ * } catch (error) {
+ *   // Handle error
+ * }
+ * ```
+ */
+export class JsonRpcEngineV2<
+  Request extends JsonRpcCall = JsonRpcCall,
+  Result extends Json = Json,
+> {
+  #middleware: Readonly<
+    NonEmptyArray<JsonRpcMiddleware<Request, Result | void>>
+  >;
+
+  readonly #makeMiddlewareIterator = (): Iterator<
+    JsonRpcMiddleware<Request, Result | void>
+  > => this.#middleware[Symbol.iterator]();
+
+  #isDestroyed = false;
+
+  constructor({ middleware }: Options<Request, Result>) {
+    this.#middleware = [...middleware];
+  }
+
+  /**
+   * Handle a JSON-RPC request. A result will be returned.
+   *
+   * @param request - The JSON-RPC request to handle.
+   * @param options - The options for the handle operation.
+   * @param options.context - The context to pass to the middleware.
+   * @returns The JSON-RPC response.
+   */
+  async handle(
+    request: Request & JsonRpcRequest,
+    options?: HandleOptions,
+  ): Promise<Result>;
+
+  /**
+   * Handle a JSON-RPC notification. No result will be returned.
+   *
+   * @param notification - The JSON-RPC notification to handle.
+   * @param options - The options for the handle operation.
+   * @param options.context - The context to pass to the middleware.
+   */
+  async handle(
+    notification: Request & JsonRpcNotification,
+    options?: HandleOptions,
+  ): Promise<void>;
+
+  async handle(
+    request: Request,
+    { context }: HandleOptions = {},
+  ): Promise<Result | void> {
+    const isReq = isRequest(request);
+    const { result } = await this.#handle(request, context);
+
+    if (isReq && result === undefined) {
+      throw new JsonRpcEngineError(
+        `Nothing ended request: ${stringify(request)}`,
+      );
+    }
+    return result;
+  }
+
+  // This exists because a JsonRpcCall overload of handle() cannot coexist with
+  // the other overloads due to type union / overload shenanigans.
+  /**
+   * Handle a JSON-RPC call. A response will be returned if the call is a request.
+   *
+   * @param request - The JSON-RPC call to handle.
+   * @param options - The options for the handle operation.
+   * @param options.context - The context to pass to the middleware.
+   * @returns The JSON-RPC response, if any.
+   */
+  async handleAny(
+    request: JsonRpcCall & Request,
+    options?: HandleOptions,
+  ): Promise<Result | void> {
+    return this.handle(request, options);
```

Comment:
nit: `await` lets us avoid unnecessarily creating a nested promise.

```suggestion
  ) {
    return (await this.handle(request, options)) satisfies Result | void;
```

Reply:
This and some other suggestions would violate `@typescript-eslint/explicit-function-return-type`, which is currently disabled for this monorepo, but there's a TODO to enable it. I understand it to be part of our convention. Is the only way to benefit from `satisfies` to remove the return annotation?

Reply:
Explicit return type annotations overwrite narrower literal types that are inferred at the call site, resulting in a loss of type information. `satisfies` is a strictly superior alternative, as it also enforces an upper bound for type safety, but doesn't prevent type inference.

> Is the only way to benefit from `satisfies` to remove the return annotation?

Unfortunately, yes. Because of this, I think we should explicitly ban `@typescript-eslint/explicit-function-return-type`. I'm drafting an entry in our contributor-docs that will reverse the [current recommendation to use explicit return annotations for functions](https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#for-functions-and-methods-provide-explicit-return-types).

RE readability concerns, the jsdoc `@returns` tag can be used for any information that needs to be included in-line or via intellisense. This also appears to be a strictly superior alternative to return type annotations.

TLDR: Explicit return type annotations contradict the following:

> **Avoid unintentionally widening an inferred type with a type annotation**
> Enforcing a wider type defeats the purpose of adding an explicit type declaration, as it loses type information instead of adding it. 
> -- https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#avoid-unintentionally-widening-an-inferred-type-with-a-type-annotation

> **Prefer `satisfies` annotation over `:` annotation for enforcing type constraints**
> the satisfies operator can be used to enforce a type constraint, while also allowing the compiler to fully narrow the assigned type through inference.
> -- https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#prefer-satisfies-annotation-over--annotation-for-enforcing-type-constraints